man/man1/pushslack.1
.\" Written by Kurt Lust, kurt.lust@uantwerpen.be for the LUMI consortium.
.TH man 1 "1 June 2026" "0.2" "pushslack command"

.SH NAME
\fBpushslack\fR \- Send notifications so Slack.

.SH DESCRIPTION
\fBpushslack\fR is a simple tool to push messages to tSlack in a safe way.
Credentials can be stored in a configuration file so that they cannot be seen as
arguments of a command in the process list. The tool uses libcurl to compose and 
send the message using the https protocol.

The tool can be configured via a configuration file. The name of that configuration
file is determined at compile time and can be found using \fBpushslack --help\fR and
is stored in your home directory. It is best to set the permissions to 600, i.e.,
group and others have no access to the file, as it is the best place to store the
credentials that are needed to send a message.

This tool supports two communication modes with Slack, though both require setting
up a Slack app. The primary mode is communication via `chat.postMessage`, while the
fallback method is via a webhook. In both cases a Slack app has to be created.
In the case of communication via `chat.postMessage`, the messages will appear in 
the "Apps" section in the workspace under the name of the app while in the 
webhook case, messages appear under your own name in "Direct messages". Even though
the webhook communication is easier to implement, this may not be desirable as you 
may want to use messages to yourself as some kind of note taking also.

To set up for `chat.postMessage` communication method:

.IP \[bu] 2
If you don't have a suitable Slack workspace yet, create one via the Slack web app
https://app.slack.com.

.IP \[bu] 2
Create a Slack app from scratch via the Slack API dashboard at https://api.slack.com/apps.
Create that app in the workspace where you want to receive your messages.

.IP \[bu] 2
In the app, under "OAuth and Permissions", scroll down the page to "Scopes" and add the 
"Bot Token Scope" "chat:write".

.IP \[bu] 2
Now you have to install or re-install the app in the workspace you created which can also be 
done from the "OAuth and Permissions" screen, a little higher up, and in "Channel for webhook", 
select your app under "Direct Messages".

Now you can find the "Bot User OAuth Token" (starts with \fBxoxb-\fR) which you will need to send 
messages. This is the argument for \fBbot-token\fR in the configuration file or the 
\fB--bot-token\fR command line option (see below).

.IP \[bu] 2
You also need to find your slack member ID in the workplace: Click on your profile picture in Slack. 
Select "Profile" in the menu. Then in the right screen with your profile, click "More" 
(the three vertical dots) and finally "Copy member ID". It will likely start with U.
This is the argument for \fBmember-id\fR in the configuration file or the
\fB--member-id\fR command line option (see below).

.PP
To set up for the webhook communication method:

.IP \[bu] 2
If you don't have a suitable Slack workspace yet, create one via the Slack web app
https://app.slack.com.

.IP \[bu] 2
Create a Slack app from scratch via the Slack API dashboard at https://api.slack.com/apps.
Create that app in the workspace where you want to receive your messages.

.IP \[bu] 2
Enable Webhooks: In the app settings, click "Incoming Webhooks" and toggle it to "On."

.IP \[bu] 2
Add Webhook to Workspace: Click "Add New Webhook" at the bottom of the screen from the previous bullet. 
When prompted to choose a channel, select your own name (the "Direct Messages" section).

.IP \[bu] 2
Copy the URL: It will look like \fBhttps://hooks.slack.com/services/T.../B.../XXXX/fR.
This is the argument for \fBwebhook-url\fR in the configuration file or the
\fB--webhook-url\fR command line option (see below).

.PP
The configuration file has the following format:

.EX
; This is a comment line
[slack]

webhook-url = https://hooks.slack.com/services/T.../B.../XXXX
bot-token = XXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
member-id = XXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
message = My default message
.EE

All key-value pairs are optional, but we highly recommend to store webhook-url,
bot-token and member-id in a configuration file and not specify them on the 
command line as then they can be intercepted
by other users and used to spam you with notifications.
You need either \fBbot-token\fR and \fBmember-id\fR or \fBwebhook-url\fR.
If both are specified, \fBpushslack\fR will use the chat.postMessage method.

.SH OPTIONS
.TP
\fB-b\fR, \fB--bot-token\fR
Specify the bot token for the chat.postMessage communication method 
on the command line, overwriting what may be set in the
configuration file. It is not recommended to use this option as the bot token can
appear in the process list if suitable options are used to see running processes.
.TP
\fB-c\fR, \fB--configuration-file\fR
Specify the configuration file to use instead of the default one.
.TP
\fB-d\fR, \fB--debug\fR
Print additional debug information.
.TP
\fB-h\fR, \fB--help\fR
Print help information about pushslack.
.TP
\fB-m\fR, \fB--message\fR
Specifies the message to be sent. This argument is optional. If a text string
is specified without flag, it will be taken as the message.
.TP
\fB-n\fR, \fB--no-configuration-file\fR
Do not read the configuration file, but use only command line arguments.
This is useful for a quick try if you have not yet created a configuration file.
.TP
\fB-u\fR, \fB--member-id\fR
Specify the memberID for the chat.postMessage communication method,
overwriting what is in the configuration file.
.TP
\fB-v\fR, \fB--version\fR
Show the version of pushover and quit.
.TP
\fB-w\fR, \fB--webhook-url\fR
Specify the webhook URL to use for the webhook communication method, overwriting
what may be set in the configuration file.
It is not recommended to use this option as the bot token can
appear in the process list if suitable options are used to see running processes.

.SH AUTHOR
Kurt Lust, University of Antwerpen, for the LUMI consortium.


src/pushslack.c
// C code to push messages to the pushover service without exposing the application key
// and user key on the screen or in process logs, etc.
//
// Written by Kurt Lust
//

#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <getopt.h>
#include "ini.h"
#include <curl/curl.h>

#define ERROR_READ_INI_FILE  1
#define ERROR_NO_BOTTOKEN    2
#define ERROR_NO_MEMBERID    3
#define ERROR_NO_WEBHOOKURL  4
#define ERROR_NO_ENV_HOME    5
#define ERROR_OUT_OF_MEMORY  6
#define ERROR_WRONG_ARGUMENT 7
#define ERROR_INTERNAL_ERROR 8

#define MODE_NONE            0
#define MODE_CHATPOSTMESSAGE 1
#define MODE_WEBHOOK         2

const char *version = "0.1";
const char *default_config_file = ".LUMI-notifications.ini";
const char *default_message = "Hello from Slack in LUMI-notifications";

typedef struct
{
    const char *bottoken;
    const char *memberid;
    const char *webhookurl;
    const char *message;
} pushslack_configuration;


//
// Print the help information.
//

void PrintHelp( void ) {

    fprintf( stderr,
        "\n"
        "Command line options for pushover:\n"
        "   -b/--bot-token              Specify the bot token, overwriting what is in the\n"
        "                               configuration file. This has precedence over \n"
        "                               --webhook-url.\n"
        "   -c/--configuration-file     Specify the configuration file to use. The default\n"
        "                               is ~/.LUMI-notifications.ini.\n"
        "   -d/--debug                  Print additional debug info.\n"
        "   -h/--help                   Print this help information and quit\n"
        "   -m/--message                Message to send.\n"
        "   -n/--no-configuration-file  Do not read any configuration file, use command\n"
        "                               line arguments instead.\n"
        "   -u/--member-id              Specify the member ID for slack, overwriting what\n"
        "                               is in the configuration file.\n"
        "   -v/--version                Print the version of pushover.\n"
        "   -w/--webhook-url            Webhook URL, overwriting what is in the configuration\n"
        "                               file. Only use if --bot-token and --member-id are not\n"
        "                               given on either the command line or in the configuration\n"
        "                               file.\n"
        "\n"
        "Default configuration file: ~/%s\n\n",
        default_config_file
    );

    return;

}  // End PrintHelp

//
// Handler for the ini tool that we use.
//

static int handler( void* user, const char* section, const char* name,
                    const char* value)
{
    pushslack_configuration* pconfig = (pushslack_configuration*) user;

    #define MATCH(s, n) strcmp(section, s) == 0 && strcmp(name, n) == 0
    if  (MATCH( "slack", "bot-token" ) ) {
        pconfig->bottoken = strdup( value );
    } else if ( MATCH( "slack", "member-id" ) ) {
        pconfig->memberid = strdup( value );
    } else if ( MATCH( "slack", "webhook-url" ) ) {
        pconfig->webhookurl = strdup( value );
    } else if ( MATCH( "slack", "message" ) ) {
        pconfig->message = strdup( value );
    } else {
        return 0;  /* unknown section/name, error */
    }
    return 1;
}


int main( int argc, char *argv[] ) {

    //
    // Get the command line arguments
    //

    const char *arg_bottoken = (char *) NULL;
    const char *arg_memberid = (char *) NULL;
    const char *arg_webhookurl = (char *) NULL;
    const char *arg_config_file = (char *) NULL;
    const char *arg_message = (char *) NULL;
    int arg_debug = 0;
    int arg_no_configuration_file = 0;

    // Define the long options
    // { "name", has_arg, *flag, val }
    static struct option long_options[] = {
        {"bot-token",             required_argument, 0, 'b'},
        {"configuration-file",    required_argument, 0, 'c'},
        {"debug",                 no_argument,       0, 'd'},
        {"help",                  no_argument,       0, 'h'},
        {"message",               required_argument, 0, 'm'},
        {"no-configuration-file", no_argument,       0, 'n'},
        {"member-id",             required_argument, 0, 'u'},
        {"version",               no_argument,       0, 'v'},
        {"webhook-url",           required_argument, 0, 'w'},
        {0,                       0,                 0, 0} // Required end of array
    };

    int option_index = 0;
    int opt;
    while ( ( opt = getopt_long( argc, argv, "b:c:dhm:nu:vw:", long_options, &option_index) ) != -1 ) {
        switch (opt) {
            case 'b':
                arg_bottoken = optarg;
                break;
            case 'c': 
                arg_config_file = optarg;
                arg_no_configuration_file = 0;
                break;
            case 'd':
                arg_debug = 1;
                break;
            case 'h': 
                PrintHelp();
                return 0;
            case 'm':
                arg_message = optarg;
                break;
            case 'n':
                arg_no_configuration_file = 1;
                break;
            case 'u':
                arg_memberid = optarg;
                break;
            case 'v':
                printf( "pushover version %s", version );
                return 0;
            case 'w':
                arg_webhookurl = optarg;
                break;
            case '?':
                if ( (optopt == 'a') || (optopt == 'c') || (optopt == 's')  || (optopt == 't') || (optopt == 'u') )
                    fprintf( stderr, "Option -%c requires an argument.\n", optopt) ;
                else if ( isprint( optopt ) )
                    fprintf( stderr, "Unknown option `-%c'.\n", optopt );
                else
                    fprintf( stderr, "Unknown option character `\\x%x'.\n", optopt);
                return ERROR_WRONG_ARGUMENT;
        } // End switch ( opt )
    } // End while

    if ( arg_message == NULL ) {
        // No message specified with -m
        if ( optind == (argc - 1) ) {
            arg_message = argv[optind];
        } else if ( optind < (argc - 1) ) {
            fprintf( stderr, "Too many arguments that may be a message." );
            return ERROR_WRONG_ARGUMENT;
        }
    } else {
        // Message already specified with -m
        if ( optind < argc ) {
            fprintf( stderr, "Too many arguments that may be a message." );
            return ERROR_WRONG_ARGUMENT;
        }
    }

    if ( arg_debug != 0 ) {
        printf( "\nValues after processing the argument list:\n" );
        if ( arg_bottoken != NULL )    printf( "- Bot token from argument list: %s\n",   arg_bottoken );
        if ( arg_memberid != NULL )    printf( "- Member ID from argument list: %s\n",   arg_memberid );
        if ( arg_webhookurl != NULL )  printf( "- Webhook URL from argument list: %s\n", arg_webhookurl );
        if ( arg_message != NULL )     printf( "- message from argument list: %s\n",     arg_message );
        if ( arg_config_file != NULL ) printf( "- config file from argument list: %s\n", arg_config_file );
        printf( "\n" );
    }


    // Read data from the INI file.

    pushslack_configuration ini_config;
    ini_config.bottoken    = NULL;
    ini_config.memberid    = NULL;
    ini_config.webhookurl  = NULL;
    ini_config.message     = NULL;
    char *config_path_file = NULL;

    if ( arg_no_configuration_file == 0 ) {

        // First compute the full path the home directory of the user.

        if ( arg_config_file == NULL ) { // No -c argument given, so the default is used.

            char *home_dir = getenv( "HOME" );
            if ( home_dir == NULL ) {
                fprintf( stderr, "Cannot locate the home directory, the environment variable HOME is missing.\n" );
                return ERROR_NO_ENV_HOME;
            }

            const int required_string_length = strlen( home_dir ) + strlen( default_config_file ) + 2; // +2: slash and terminating null

            config_path_file = (char *) malloc( (size_t) (required_string_length * sizeof( char )) );
            if ( config_path_file == NULL ) {
                fprintf( stderr, "Memory allocation failed in file %s at line %d.\n", __FILE__, __LINE__ );
                return ERROR_OUT_OF_MEMORY;
            }

            snprintf( config_path_file, required_string_length, "%s/%s", 
                      home_dir, default_config_file );

        } else
            config_path_file = (char *) arg_config_file;

        // Now read the actual data from the configuration file.

        if ( ini_parse( config_path_file, handler, &ini_config ) < 0 ) {
            fprintf( stderr, "Failed to load the INI file %s\n", config_path_file );
            return ERROR_READ_INI_FILE;
        }

        if ( arg_debug != 0 ) {
            printf( "Values read from the configuration file %s:\n", config_path_file );
            if ( ini_config.bottoken != NULL )   printf( "- Bot token from argument list: %s\n",   ini_config.bottoken );
            if ( ini_config.memberid != NULL )   printf( "- Member ID from argument list: %s\n",   ini_config.memberid );
            if ( ini_config.webhookurl != NULL ) printf( "- Webhook URL from argument list: %s\n", ini_config.webhookurl );
            if ( ini_config.message != NULL )    printf( "- message from argument list: %s\n",     ini_config.message );
            printf( "\n" );
        }

    } else if ( arg_debug != 0 )
        printf( "No configuration file read.\n\n" );

    pushslack_configuration config; // Actual configuration to use to send the message

    int source_bottoken = 0;   // Variable used to track the source of the API token for debug output
    int source_memberid = 0;   // Variable used to track the source of the user key for debug output
    int source_webhookurl = 0; // Variable used to track the source of the title for debug output
    int source_message = 0;    // Variable used to track the source of the message for debug output

    // Determine the actual bot token (if any)

    if ( arg_bottoken != NULL ) {
        config.bottoken = arg_bottoken;
        source_bottoken = 1;
    } else if ( ini_config.bottoken != NULL ) {
        config.bottoken = ini_config.bottoken;
        source_bottoken = 2;
    } else {
        config.bottoken = NULL;
        source_bottoken = 0;
    }
    // Determine the actual memberid (if any)

    if ( arg_memberid != NULL ) {
        config.memberid = arg_memberid;
        source_memberid = 1;
    } else if ( ini_config.memberid != NULL ) {
        config.memberid = ini_config.memberid;
        source_memberid = 2;
    } else {
        config.memberid = NULL;
        source_memberid = 0;
    }
   
    // Determine the webhook URL (if any)

    if ( arg_webhookurl != NULL ) {
        config.webhookurl = arg_webhookurl;
        source_webhookurl = 1;
    } else if ( ini_config.webhookurl != NULL ) {
        config.webhookurl = ini_config.webhookurl;
        source_webhookurl = 2;
    } else {
        config.webhookurl = NULL;
        source_webhookurl = 0;
    }
   
    // Determine the message

    if ( arg_message != NULL ) {
        config.message = arg_message;
        source_message = 1;
    } else if ( ini_config.message != NULL ) {
        config.message = ini_config.message;
        source_message = 2;
    } else {
        config.message = default_message;
        source_message = 0;
    }

    if ( arg_debug != 0 ) {
        const char *source[] = { "default", "command line", "configuration file"};
        printf( "Combined parameters from command line and ini file:\n" );
        if ( config.bottoken != NULL )   printf( "- bot token: %s (%s)\n",   config.bottoken,   source[source_bottoken] );
        else                             printf( "- bot token not specified.\n" );
        if ( config.memberid != NULL )   printf( "- Member ID: %s (%s)\n",   config.memberid,   source[source_memberid] );
        else                             printf( "- Member ID not specified.\n" );
        if ( config.webhookurl != NULL ) printf( "- Webhook URL: %s (%s)\n", config.webhookurl, source[source_webhookurl] );
        else                             printf( "- Webhook URL not specified\n" );
        if ( config.message != NULL )    printf( "- message: %s (%s)\n",     config.message,    source[source_message] );
        printf( "\n" );
    }

    // Now determine the communication mode: 1 = chat.postMessage, 2 = webhook

    int communication_mode = MODE_NONE;

    if ( ( config.bottoken != NULL ) && ( config.memberid != NULL ) )
        communication_mode = MODE_CHATPOSTMESSAGE;
    else if ( ( config.bottoken != NULL ) || ( config.memberid != NULL ) ) {
        // Either config.bottoken is not NULL or memberid is not NULL but not both
        if ( config.bottoken == NULL ) {
            fprintf( stderr, "No bot token given even though the member ID was given. Either both or none should be given.\n" );
            fprintf( stderr, "The bot token should be given either in %s (field bot-token in [slack]) or on the command line (--bot-token).\n", config_path_file );
            if ( source_memberid == 1 ) fprintf( stderr, "The member ID was taken from the command line\n" );
            else                        fprintf( stderr, "The member ID was found in the configuration file %s.\n", config_path_file );
            return ERROR_NO_BOTTOKEN;
        } else {
            fprintf( stderr, "No member ID given even though the bot token was given. Either both or none should be given.\n" );
            fprintf( stderr, "The member id should be given either in %s (field member-id in [slack]) or on the command line (--member-id).\n", config_path_file );
            if ( source_memberid == 1 ) fprintf( stderr, "The member ID was taken from the command line\n" );
            else                        fprintf( stderr, "The member id was found in the configuration file %s.\n", config_path_file );
            return ERROR_NO_MEMBERID;
        }
    } else if ( config.webhookurl != NULL ) {
        communication_mode = MODE_WEBHOOK;
    } else {
        fprintf( stderr, "Either a bot token and member-id need to be specified, or a webhook URL, but neither were found.\n" );
        return ERROR_NO_WEBHOOKURL;
    }

    if ( arg_debug != 0 ) {
        switch ( communication_mode ) {
            case MODE_CHATPOSTMESSAGE : printf( "Communicating via chat.postMessage.\n\n" ); break;
            case MODE_WEBHOOK         : printf( "Communicating via a webhook URL.\n\n" ); break;
            default :                   fprintf( stderr, "Should never have arrived at this point. File %s at line %d.\n", __FILE__, __LINE__ ); return ERROR_INTERNAL_ERROR;
        }
    }

    // Now use libcurl to push a message.

    CURL *curl;
    CURLcode curl_res;

    curl_global_init( CURL_GLOBAL_ALL );
    curl = curl_easy_init();
    if ( curl == (CURL *) NULL ) {
        fprintf( stderr, "Failed to initialise libcurl properly to send a message in file %s at line %d.\n", __FILE__, __LINE__ );
        return ERROR_INTERNAL_ERROR;
    }

    if( curl ) {
    
        if ( communication_mode == MODE_CHATPOSTMESSAGE ) { // Communicating via chat.postMessage

            struct curl_slist *headers = NULL;

            int required;

            // Header: Authorization: Bearer token
            size_t token_length = 23 + strlen( config.bottoken );
            char *token = (char *) malloc( (size_t) (token_length * sizeof( char )) );
            if ( token == (char *) NULL ) { fprintf( stderr, "Out-of-memory in file %s at line %d.\n", __FILE__, __LINE__ ); return ERROR_OUT_OF_MEMORY; }
            required = snprintf( token, token_length, "Authorization: Bearer %s", config.bottoken );
            if ( required >= (int) token_length ) { fprintf( stderr, "Internal error in file %s at line %d.\n", __FILE__, __LINE__ ); return ERROR_INTERNAL_ERROR; }
            headers = curl_slist_append(headers, token );

            // Header: Content-Type: application/json
            headers = curl_slist_append(headers, "Content-Type: application/json");

            // Make the json message.
            const int bufsize = 25; // Large enough that Cray Clang does not complain that the snprintf buffer will always be truncated.
            char dummybuf[bufsize];
            int json_length = snprintf( dummybuf, (size_t) bufsize, "{\"channel\":\"%s\",\"text\":\"%s\"}", config.memberid, config.message ) + 1;
            char *json = (char *) malloc( (size_t) (json_length * sizeof( char )) );
            if ( json == (char *) NULL ) { fprintf( stderr, "Out-of-memory in file %s at line %d.\n", __FILE__, __LINE__ ); return ERROR_OUT_OF_MEMORY; }
            required = snprintf( json, (size_t) json_length, "{\"channel\":\"%s\",\"text\":\"%s\"}", config.memberid, config.message );
            if ( required >= (int) json_length ) { fprintf( stderr, "Internal error in file %s at line %d.\n", __FILE__, __LINE__ ); return ERROR_INTERNAL_ERROR; }

            curl_easy_setopt( curl, CURLOPT_URL,        "https://slack.com/api/chat.postMessage" );
            curl_easy_setopt( curl, CURLOPT_HTTPHEADER, headers );
            curl_easy_setopt( curl, CURLOPT_POSTFIELDS, json );

            // We'll send the output to /dev/null if we are not debugging to avoid bombarding the user
            // with output they cannot understand.
            FILE *devnull;
            if ( arg_debug == 0 ) {
                devnull = fopen( "/dev/null", "w" );
                curl_easy_setopt( curl, CURLOPT_WRITEDATA, devnull );
            }

            // Perform the request
            if ( arg_debug != 0 ) printf( "Calling curl, returns: " );
            curl_res = curl_easy_perform( curl );
            if ( arg_debug != 0 ) printf( "\n\n" );

            if( curl_res != CURLE_OK ) {
                fprintf( stderr, "Sending the message failed, cURL returned: %s\n", curl_easy_strerror( curl_res ) );
            } else {
                printf( "Message sent successfully!\n" );
            }

            // Clean-up
            curl_slist_free_all( headers );
            curl_easy_cleanup( curl );

            free( token );
            free( json );

            if ( arg_debug == 0 ) fclose( devnull );

        } else { // Communicating via webhook URL

            struct curl_slist *headers = NULL;

            int required;

            // Build the header, in this case only the Content-Type: application/json
            headers = curl_slist_append( headers, "Content-Type: application/json" );

            // Build the json part
            const int bufsize = 12; // Large enough that Clang does not complain that the snprintf buffer will always be truncated.
            char dummybuf[bufsize];
            int json_length = snprintf( dummybuf, (size_t) bufsize, "{\"text\":\"%s\"}", config.message ) + 1;
            char *json = (char *) malloc( (size_t) (json_length * sizeof( char )) );
            if ( json == (char *) NULL ) { fprintf( stderr, "Out-of-memory in file %s at line %d.\n", __FILE__, __LINE__ ); return ERROR_OUT_OF_MEMORY; }
            required = snprintf( json, (size_t) json_length, "{\"text\":\"%s\"}", config.message );
            if ( required >= (int) json_length ) { fprintf( stderr, "Internal error in file %s at line %d.\n", __FILE__, __LINE__ ); return ERROR_INTERNAL_ERROR; }

            // Build the curl call
            curl_easy_setopt( curl, CURLOPT_URL,        config.webhookurl );
            curl_easy_setopt( curl, CURLOPT_HTTPHEADER, headers );
            curl_easy_setopt( curl, CURLOPT_POSTFIELDS, json );

            // We'll send the output to /dev/null if we are not debugging to avoid bombarding the user
            // with output they cannot understand.
            FILE *devnull;
            if ( arg_debug == 0 ) {
                devnull = fopen( "/dev/null", "w" );
                curl_easy_setopt( curl, CURLOPT_WRITEDATA, devnull );
            }

            // Perform the request
            if ( arg_debug != 0 ) printf( "Calling curl, returns: " );
            curl_res = curl_easy_perform( curl );
            if ( arg_debug != 0 ) printf( "\n\n" );

            if( curl_res != CURLE_OK ) {
                fprintf( stderr, "Sending the message failed, cURL returned: %s\n", curl_easy_strerror( curl_res ) );
            } else {
                printf( "Message sent successfully!\n" );
            }

            // Clean-up
            curl_slist_free_all( headers );
            curl_easy_cleanup( curl );

            free( json );

            if ( arg_debug == 0 ) fclose( devnull );

        } // end else-part if ( communication_mode ... )

    curl_global_cleanup();

    // It's a puzzle to figure out which memory must be freed, and as this is done
    // automatically when a program terminates, we don't care.

    return 0;

    } // end if ( curl )

}

